lib/flight_scheduler/message_processor.rb

```diff
@@ -140,20 +143,47 @@ def call(message)
       when 'JOB_CANCELLED'
         job_id = message[:job_id]
         Async.logger.info("Cancelling job:#{job_id}")
-        job_runner = FlightScheduler.app.job_registry.lookup_runner(job_id, 'BATCH')
-        if job_runner
-          job_runner.cancel
-          # The RUN_SCRIPT task will report back that the process has failed.
-          # We don't need to send any messages to the controller here.
+
+        # Deallocate the job to prevent any further job steps
+        FlightScheduler.app.job_registry.deallocate_job(job_id)
+
+        Async do |task|
+          # Allow other tasks to run a final time before cancelling
+          # This is a last attempt to collect any finished processes
+          task.yield
+
+          # Cancel all current runners
+          FlightScheduler.app.job_registry.lookup_runners(job_id).each do |_, runner|
+            runner.cancel
+          end
+
+          # Wait for the runners to finish and remove the job
+          task.yield until FlightScheduler.app.job_registry.lookup_runners(job_id).empty?
+          FlightScheduler.app.job_registry.remove_job(job_id)
+          MessageSender.send(command: 'NODE_DEALLOCATED', job_id: job_id)
+        end
+
+      when 'JOB_DEALLOCATED'
+        job_id = message[:job_id]
+        Async.logger.info("Deallocating job:#{job_id}")
+
+        # Deallocate the job to prevent any further job steps
+        FlightScheduler.app.job_registry.deallocate_job(job_id)
+
+        # Report back when all the runners have stop
+        Async do |task|
+          task.yield until FlightScheduler.app.job_registry.lookup_runners(job_id).empty?
```

Comment:
Again `task.sleep(1)` please.

Reply:
https://github.com/openflighthpc/flight-scheduler-daemon/pull/16/commits/7f5b1d79396a1b8f657fe9c4bd1fa3abbfc42ee1

Reply:
I've made it 0.1 for consistency with `MessageSender` wait period. Also opened a trello card to review the wait periods:

https://trello.com/c/V4haW0L6/125-make-the-various-async-sleeps-configurable